### PR TITLE
feat: create installation script for mentorship-backend

### LINKED ISSUE #223: Create an Installation script to automatically install and setup the project
**Is your feature request related to a problem? Please describe.**
Currently the installation and setup of Mentorship Backend is done manually. It would be great to have a single script which could download the code and setup the various configurations required automatically. 

**Describe the solution you'd like**
Create a shell script named **install.sh** (for Linux/Mac) or **install.bat** (for Windows) which would do the following when run - 
    - Clone the repository to the present working directory of the running script (if not already inside one, which will mostly be the case )
    - Perform all the required setup for the repository to run. Take any variable inputs as mandatory command line arguments, without which the script will produce an error, and prompt the user to enter the values. 
    - Perform unit testing on the installation and display its results

**Describe alternatives you've considered**
To modify the deploy.sh file to work on local systems

### PR DESCRIPTION
**Pull request by daksha-aeer**
### Description
Linux .sh script performs the following
1. Check if git repo, .env and virtual environment are present, otherwise clone them
2. take environment variables as input variables
3. install pip dependencies from requirements.txt
3. perform tests
4. launch backend on localhost

Fixes #223

### Type of Change:
- Code

### How Has This Been Tested?
1. **Script working**

```sh
./install.sh
```
task17_3

2. **All tests passed**
3. **Result**
task17_2

### Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+echo "export MAIL_DEFAULT_SENDER=$MAIL_DEFAULT_SENDER" >> .env
+echo "export MAIL_SERVER=$MAIL_SERVER" >> .env
+echo "export APP_MAIL_USERNAME=$APP_MAIL_USERNAME" >> .env
+echo "export APP_MAIL_PASSWORD=$APP_MAIL_PASSWORD" >> .env
+
+# create venv
+python3 -m venv venv
+source venv/bin/activate
+pip3 install -r requirements.txt
+python -m unittest discover tests
```
Comment (sara-02) [inline install.sh]: For sanity's sake, this should also be `python3`

Comment (daksha-aeer) [reply install.sh]: Done, please check

File: install.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+echo "export FLASK_ENVIRONMENT_CONFIG=$FLASK_ENVIRONMENT_CONFIG" >> .env
+echo "export SECRET_KEY=$SECRET_KEY" >> .env
+echo "export SECURITY_PASSWORD_SALT=$SECURITY_PASSWORD_SALT" >> .env
+echo "export MAIL_DEFAULT_SENDER=$MAIL_DEFAULT_SENDER" >> .env
+echo "export MAIL_SERVER=$MAIL_SERVER" >> .env
+echo "export APP_MAIL_USERNAME=$APP_MAIL_USERNAME" >> .env
+echo "export APP_MAIL_PASSWORD=$APP_MAIL_PASSWORD" >> .env
+
+# create venv
+python3 -m venv venv
```
Comment (sara-02) [inline install.sh]: Not everyone will use virtualenv via this command.
A better way is the step 1 mentioned in our `README`
https://github.com/systers/mentorship-backend#run-app
I think we should do something like
`pip3 install --user virtualenv`
and then `virtualenv --python /usr/bin/python3 venv`

Comment (sara-02) [reply install.sh]: I ran into the same trouble while installing it.

Comment (daksha-aeer) [reply install.sh]: Done please review

File: install.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+echo "export FLASK_ENVIRONMENT_CONFIG=$FLASK_ENVIRONMENT_CONFIG" >> .env
+echo "export SECRET_KEY=$SECRET_KEY" >> .env
+echo "export SECURITY_PASSWORD_SALT=$SECURITY_PASSWORD_SALT" >> .env
+echo "export MAIL_DEFAULT_SENDER=$MAIL_DEFAULT_SENDER" >> .env
+echo "export MAIL_SERVER=$MAIL_SERVER" >> .env
+echo "export APP_MAIL_USERNAME=$APP_MAIL_USERNAME" >> .env
+echo "export APP_MAIL_PASSWORD=$APP_MAIL_PASSWORD" >> .env
+
+# create venv
+pip3 install virtualenv
```
Comment (sara-02) [inline install.sh]: `pip3 --install user virtualenv` otherwise some users might run into sudo permission issues.`

Comment (daksha-aeer) [reply install.sh]: Added

**Comment by sara-02** [comment]
LGTM.
This covers the task.
For future I am just adding a comment, how can we further improve the installation process if some parts of the script fail. Like if even one of the env fails, the script exits. But I cannot run the script again without first deleting the `mentorship-backend` folder.

**Comment by sara-02** [comment]
Or how can we detach the `python run.py` command so that the script can successfully exist once the service is up and running? 
Should the script stop if one of the dependency fails to install, how to check that?

**Comment by xprilion** [comment]
@daksha-aeer I suggest you to check before taking any input if the mentorship-backend directory already exists and display a deletion warning/confirmation before proceeding. 

Also, at the end of the script you might want to display a success/error message. 

Further, you don't need to explicitly take input for the `APP_MAIL_USERNAME` value. It can be and should be generated by making a substring.

**Comment by daksha-aeer** [comment]
> @daksha-aeer I suggest you to check before taking any input if the mentorship-backend directory already exists and display a deletion warning/confirmation before proceeding.
> 
> Also, at the end of the script you might want to display a success/error message.
> 
> Further, you don't need to explicitly take input for the `APP_MAIL_USERNAME` value. It can be and should be generated by making a substring.



> @daksha-aeer I suggest you to check before taking any input if the mentorship-backend directory already exists and display a deletion warning/confirmation before proceeding.
> 
> Also, at the end of the script you might want to display a success/error message.
> 
> Further, you don't need to explicitly take input for the `APP_MAIL_USERNAME` value. It can be and should be generated by making a substring.

Ok working on it. But I had a question as well. The task asked us to create a script which clones the backend repo. But isn't the script itself bundled with the repo? Maybe as later work the cloning part in the script could be removed?

**Comment by daksha-aeer** [comment]
> @daksha-aeer I suggest you to check before taking any input if the mentorship-backend directory already exists and display a deletion warning/confirmation before proceeding.
> 
> Also, at the end of the script you might want to display a success/error message.
> 
> Further, you don't need to explicitly take input for the `APP_MAIL_USERNAME` value. It can be and should be generated by making a substring.

@xprilion Should I make a separate commit or force push the changes? For this task got approved from mentors' side.

**Comment by xprilion** [comment]
@daksha-aeer feel free to add more commits! While you've already met the minimum requirements for this task to get accepted on the GCI program side, here on GitHub, you're encouraged to work on it anytime for improving it till the end of GCI (or till it is merged) to make it better!

**Comment by sara-02** [comment]
@daksha-aeer the Systers format is just one commit per PR. So you need to make sure that it follows :)

**Comment by daksha-aeer** [comment]
Changes made, please take a look

**Comment by SanketDG** [comment]
@isabelcosta Do we need this?

**Comment by vj-codes** [comment]
Closing since issue is not assigned to the author 
Thank you for your contribution :)

**Pull request closed without merging**